### ISSUE TITLE
[Enh]: Registering extensions via narwhals to downstream dataframe libraries

### ISSUE THREAD

**Issue by FBruzzesi** [narwhals-dev/narwhals#267]
### We would like to learn about your use case. For example, if this feature is needed  to adopt Narwhals in an open source project, could you please enter the link to it below?

Direct use case I have in mind is a library I created and maintain: [iso-week-date](https://github.com/FBruzzesi/iso-week-date)

### Please describe the purpose of the new feature or describe the problem to solve.

The objective is to have a way to register an extension namespace in narwhals once, which is then propagated to the downstream libraries.

Here is the problem/headache I face: as a library creator I want to provide both a [pandas](https://pandas.pydata.org/docs/development/extending.html) and a [polars](https://docs.pola.rs/py-polars/html/reference/api.html) extension by registering namespaces.

In the current scenario, I am duplicating explicitly writing every method for each library ([pandas](https://github.com/FBruzzesi/iso-week-date/blob/main/iso_week_date/pandas_utils.py#L324) and [polars](https://github.com/FBruzzesi/iso-week-date/blob/main/iso_week_date/polars_utils.py#L328)), which is far from ideal.

### Suggest a solution if possible.

The ideal scenario would be to write the extension once using narwhals and decorate it for each downstream library.

Maybe this is already possible, but I would like to double check if that's the case (and if the answer is positive I think it would be awesome to document). Here is the closest I got:

**_core.py**
```py
import narwhals as nw

class PlusOneSeriesNW:
    def __init__(self, series: nw.Series):
        self._series: nw.Series = series

    def plus_one(self):
         return self.series + 1
```

**pandas_extension.py**
```py
from ._core import PlusOneSeriesNW

@pd.api.extensions.register_series_accessor("p1")
class PlusOneSeries:
    def __init__(self, series: pd.Series):
        self._series = PlusOneSeriesNW(nw.from_series(series, ...))
    
    def plus_one(self):
        return nw.to_series(self._series.plus_one(), ...)
```
Now if someone imports `PlusOneSeries` from the library, it will be possible to run: `series.p1.plus_one()`

**polars_extension.py**
> Ok just same as above

### Would you want to open a pull request?

yes

**Comment by wd60622** [comment]
Would be an awesome addition. I believe I have the same use-case as described for my package latent-calendar where my main API is [a pandas extension](https://wd60622.github.io/latent-calendar/examples/cal-attribute/). I would like generalize this if I could!

**Comment by MarcoGorelli** [comment]
thanks @wd60622 ! the priority at the moment is the plotly integration, but we'll get back to this shortly 👌

**Comment by FBruzzesi** [comment]
I spent ~30 minutes to check a few scenarios on this topic.

Consider the following example for a series extension:

```py
import narwhals.stable.v1 as nw
import numpy as np
import pandas as pd
import polars as pl

class NwMathShortcuts:
    """Base class, should not be used directly."""

    def __init__(self, s) -> None:
        self._s = nw.from_native(s, series_only=True)

    def square(self) -> nw.Series:
        return nw.to_native(self._s * self._s)

    def cum_sum(self) -> nw.Series:
        return nw.to_native(self._s.cum_sum())

    def alternating_sum(self) -> float:
        w = nw.new_series(
            name="weights",
            values=np.resize([1,-1], len(self._s)),
            native_namespace=nw.get_native_namespace(self._s)
        )
        return (self._s * w).sum()

@pl.api.register_series_namespace("math")
class PlMathShortcuts(NwMathShortcuts):
    pass

@pd.api.extensions.register_series_accessor("math")
class PdMathShortcuts(NwMathShortcuts):
    pass

data = list(range(1, 6))

s_pl = pl.Series("n", data)
s_pl.math.square()
s_pl.math.cum_sum()
s_pl.math.alternating_sum()

s_pd = pd.Series([1, 2, 3, 4, 5], name="n")
s_pd.math.square()
s_pd.math.cum_sum()
s_pd.math.alternating_sum()
```

The same approach would work for dataframe.
However, for polars Expr it would be a bit weird (especially in the typing system), since `nw.from_native` will pass through Expressions (I have a use case for this, i.e. for functionalities and extensions that should work for both polars series, polars expr and pandas series).

Now as you can see from the example, there is _some_ level of repetition of `from_native` and `to_native`.

We _could_ provide a metaclass to handles that: `class NwMathShortcuts(NwExtensionMeta, ...)`. However, I would rather just suggest that in a documentation tutorial, but not maintain such metaclass directly as I can imagine a lot of customization might be needed.

TL;DR:

- Using inheritance seems a quick win
- Extending polars Expr might be not ideal for the typing system
- Metaclass would simplify some boilerplate, but I suggest to make a tutorial for it instead of maintaining it directly

**Comment by MarcoGorelli** [comment]
i'm not sure we should be enabling this, i think this is the kind of things that could be built on top of narwhals
